### PR TITLE
Added multi level investment policy in Credit Mining

### PR DESCRIPTION
**Pull request by xoriole**
**Multi level Investment Policy for Credit Mining.**

Currently, uses 20 level investment states. Each investment state represents upload or download mode, threshold bandwidth limit, and promotion ratio. A torrent has to fulfill the bandwidth limit of upload or download with a promotion ratio before it can go to a higher state. This enforces that good investments rise to a higher level while bad investments ultimately get removed.

This PR is ready for review

### CODE REVIEW

**Comment by xoriole** [comment]
Credit mining update on the GUI
token_mining_update

**Comment by xoriole** [comment]
retest this please

**Comment by xoriole** [comment]
retest this please

**Comment by xoriole** [comment]
retest this please

**Comment by xoriole** [comment]
Probably unrelated error to credit mining

ERROR: Test the endpoint to fetch the health of a chant-managed, infohash-only torrent
```
ResponseNeverReceived: [<twisted.python.failure.Failure twisted.internet.error.ConnectionDone: Connection was closed cleanly.>]
-------------------- >> begin captured logging << --------------------
DEBUG   1548333201   network_utils:Tribler.Core.Utilities.network_utils:38   Got a working random port 10911
INFO    1548333201         Session:Session:75   Session is using state directory: C:\workspace\jenkins2\workspace\GH_Tribler_PR_Tests\PR_win64\tribler\Tribler\Test\temp\TestTorrentHealthEndpoint_tribler_test_session_BGZYWJ\dot.Tribler
DEBUG   1548333201   network_utils:Tribler.Core.Utilities.network_utils:38   Got a working random port 59204
DEBUG   1548333201  tribler_config:TriblerConfig:124  Get random port 59204 for [libtorrent~port]
DEBUG   1548333201   network_utils:Tribler.Core.Utilities.network_utils:38   Got a working random port 40008
DEBUG   1548333201  tribler_config:TriblerConfig:124  Get random port 40008 for [dispersy~port]
DEBUG   1548333201   network_utils:Tribler.Core.Utilities.network_utils:38   Got a working random port 21403
DEBUG   1548333201  tribler_config:TriblerConfig:124  Get random port 21403 for [mainline_dht~port]
DEBUG   1548333201   network_utils:Tribler.Core.Utilities.network_utils:38   Got a working random port 40723
DEBUG   1548333201  tribler_config:TriblerConfig:124  Get random port 40723 for [video_server~port]
DEBUG   1548333201   network_utils:Tribler.Core.Utilities.network_utils:38   Got a working random port 59999
DEBUG   1548333201  tribler_config:TriblerConfig:124  Get random port 59999 for [libtorrent~anon_listen_port]
DEBUG   1548333201   network_utils:Tribler.Core.Utilities.network_utils:38   Got a working random port 56039
DEBUG   1548333201  tribler_config:TriblerConfig:124  Get random port 56039 for [tunnel_community~socks5_listen_ports~0]
DEBUG   1548333201   network_utils:Tribler.Core.Utilities.network_utils:38   Got a working random port 17950
DEBUG   1548333201  tribler_config:TriblerConfig:124  Get random port 17950 for [tunnel_community~socks5_listen_ports~1]
DEBUG   1548333201   network_utils:Tribler.Core.Utilities.network_utils:38   Got a working random port 49278
DEBUG   1548333201  tribler_config:TriblerConfig:124  Get random port 49278 for [tunnel_community~socks5_listen_ports~2]
DEBUG   1548333201   network_utils:Tribler.Core.Utilities.network_utils:38   Got a working random port 45034
DEBUG   1548333201  tribler_config:TriblerConfig:124  Get random port 45034 for [tunnel_community~socks5_listen_ports~3]
DEBUG   1548333201   network_utils:Tribler.Core.Utilities.network_utils:38   Got a working random port 5988
DEBUG   1548333201  tribler_config:TriblerConfig:124  Get random port 5988 for [tunnel_community~socks5_listen_ports~4]
INFO    1548333202   sqlitecachedb:SQLiteCacheDB:144  begin page_size upgrade...
INFO    1548333202   sqlitecachedb:SQLiteCacheDB:148  ...end page_size upgrade
INFO    1548333202   sqlitecachedb:SQLiteCacheDB:168  Initializing new database...
INFO    1548333202   sqlitecachedb:SQLiteCacheDB:189  Reading database version...
INFO    1548333202   sqlitecachedb:SQLiteCacheDB:193  Current database version is 28
INFO    1548333202   sqlitecachedb:SQLiteCacheDB:217  Beginning the first transaction...
DEBUG   1548333202  LaunchManyCore:TriblerLaunchMany:165  tlm: Reading Session state from C:\workspace\jenkins2\workspace\GH_Tribler_PR_Tests\PR_win64\tribler\Tribler\Test\temp\TestTorrentHealthEndpoint_tribler_test_session_BGZYWJ\dot.Tribler
DEBUG   1548333202    FamilyFilter:XXXFilter:44   Read 3902 XXX terms from file C:\workspace\jenkins2\workspace\GH_Tribler_PR_Tests\PR_win64\tribler\Tribler\..\Tribler\Core\Category\filter_terms.filter
DEBUG   1548333202        Category:Category:37   category: Categories defined by user: [('Video', 'Video Files'), ('VideoClips', 'Video Clips'), ('Audio', 'Audio'), ('Compressed', 'Compressed'), ('Document', 'Documents'), ('Picture', 'Pictures'), ('other', 'Other'), ('unknown', 'Unknown'), ('xxx', 'XXX')]
DEBUG   1548333202 SqliteCacheDBHandler:ChannelCastDBHandler:1415 Channels: my channel is None
DEBUG   1548333202  LaunchManyCore:TriblerLaunchMany:779  Starting the download state callback with interval 1.000000
DEBUG   1548333202   network_utils:Tribler.Core.Utilities.network_utils:38   Got a working random port 10581
DEBUG   1548333202   network_utils:Tribler.Core.Utilities.network_utils:38   Got a working random port 11732
DEBUG   1548333202 torrent_checker:TorrentChecker:108  tracker selection interval changed to 20
```

**Comment by xoriole** [comment]
retest this please

File: Tribler/Core/CreditMining/CreditMiningPolicy.py [hunk 1]
```diff
@@ -51,3 +114,178 @@ def sort_key(torrent):
...
+                          self.started_in_iteration, self.stopped_in_iteration)
+        self._logger.info('Torrents in upload mode: %d, download mode: %d',
+                          self.num_uploading_in_iteration, self.num_downloading_in_iteration)
+
+    def get_reserved_bytes(self, torrent):
+        """
+        Returns the storage(bytes) required for the torrent to move to next investment state's bandwidth limit.
+        """
+        investment_state = self.investment_states[torrent.mining_state.get('state_id', 0)]
+        _, downloaed_bytes = torrent.get_storage()
```
Comment (devos50) [inline Tribler/Core/CreditMining/CreditMiningPolicy.py]: There's a typo in this variable name.

File: Tribler/Core/Libtorrent/LibtorrentDownloadImpl.py [hunk 2]
```diff
@@ -1032,12 +1040,13 @@ def restart(self):
                 def schedule_create_engine(_):
                     self.cew_scheduled = True
                     create_engine_wrapper_deferred = self.network_create_engine_wrapper(
-                        self.pstate_for_restart, share_mode=self.get_share_mode())
+                        self.pstate_for_restart, share_mode=self.get_share_mode(), upload_mode=upload_mode)
```
Comment (devos50) [inline Tribler/Core/Libtorrent/LibtorrentDownloadImpl.py]: Could you make `upload_mode` a config option of the download?

**Pull request merged**